content/reference/forms.md

```diff
@@ -0,0 +1,311 @@
+app.route('/', main)
+app.mount('body')
+
+function main () {
+  return html`
+    <body>
+      <form id="login" onsubmit=${onsubmit}>
+        <label for="username">
+          username
+        </label>
+        <input id="username" name="username"
+          type="text"
+          required
+          pattern=".{1,36}"
+          title="Username must be between 1 and 36 characters long."
+        >
+        <label for="password">
+          password
+        </label>
+        <input id="password" name="password"
+          type="password"
+          required
+        >
+        <input type="submit" value="Login">
+      </form>
+    </body>
+  `
+
+  function onsubmit (e) {                                               // 1.
+    e.preventDefault()
+    var form = document.querySelector('#login')
+    var data = new FormData(form)                                       // 2.
+    var headers = new Headers({ 'Content-Type': 'application/json' })   // 3.
+    var body = {}
+    for (var pair of data.entries()) body[pair[0]] = pair[1]            // 4.
+    body = JSON.stringify(body)                                         // 5.
+    fetch('/dashboard', { method: 'POST', body, headers })              // 6.
+      .then(res => {
+        if (!res.ok) return console.log('oh no!')
+        console.log('request ok \o/')
+      })
+      .catch(err => console.log('oh no!'))
+}
+```
+
+1. We create a handler for the `'submit'` event. This will fire whenever a user
+   clicks the `type="submit"` button (or an equivalent action).
+2. We select the `<form>` element, and extract all of its data into a
+   `FormData` instance.
+3. We need to send data as `application/json`, so we create a new `Headers`
+   object that we can later attach to our `fetch()` call.
+4. We need to convert the `FormData` instance to an `Object`. This means
+   iterating over it, and copying each key-value pair.
+5. Now that we have a regular `Object`, we can convert it into JSON using
+   `JSON.stringify`.
+6. With our `body` and `headers` ready, we can send a `POST` request down to a
+   server.
+
+_Note: perhaps you're thinking to yourself this might be a lot of typing, and
+you wouldn't be wrong! We wanted to show you what it's like to make requests
+using only DOM APIs. If you're planning to use this to write applications, it's
+probably best to use small abstractions to `POST` data, and convert `<form>`
+elements into `JSON`._
+
+## Uploading files
+This is a hard thing to write about. Uploading files is similar to
+`multipart/form-data`, but usually requires some extra features - such as
+overriding the native form controls, showing upload progress, and validation
+strategies. There's quite a bit to cover here.
+
+Instead of covering everything, we're going to share a few useful snippets.
+Because of time constraints, we can't quite write a full section about this yet.
+But we hope this is enough to help you on your way. Pull requests to add to
+this would be amazing!
```

Comment:
perhaps link to how to contribute to this document?

Reply:
There should be an edit button at the top when we put it on the website. Changed the phrasing a little tho; hopefully that helps! (: